server/tokenRegistryProxy.js

```diff
@@ -1,29 +1,95 @@
 require('isomorphic-fetch')
 const Sentry = require('@sentry/node')
 const {backendConfig} = require('./helpers/loadConfig')
-const cacheResults = require('./helpers/cacheResults')
+const chunk = require('./helpers/chunk')
+const Cache = require('./helpers/cache')
 
-const CACHE_TIMEOUT = 24 * 60 * 60 * 1000 // 1 day
+const REQUEST_CHUNK_SIZE = 100
+const MAX_REQUEST_SIZE = 2000
 
-const cachedFetch = cacheResults(CACHE_TIMEOUT)(async (url) => {
-  const response = await fetch(url)
-  if (response.status === 200) {
-    return response.json()
-  } else {
-    return null
+const cache = (() => {
+  // We expect far more tokens to not be in token registry, we keep that information in nullCache
+  // Keeping it separate helps us to not prune cached information about tokens that have real
+  // metadata in token registry
+  const cache = new Cache(1000)
+  const nullCache = new Cache(100000)
+
+  const get = (key) => {
+    const value = cache.get(key)
+    if (value === undefined) return nullCache.get(key)
+    return value
+  }
+
+  const set = (key, value) => {
+    if (value === null) {
+      nullCache.set(key, value)
+    } else {
+      cache.set(key, value)
+    }
   }
-})
+
+  return {get, set}
+})()
 
 module.exports = function(app, env) {
   app.post('/api/bulk/tokens/metadata', async (req, res) => {
     try {
-      const responses = await Promise.all(
-        req.body.subjects.map((subject) =>
-          cachedFetch(`${backendConfig.ADALITE_TOKEN_REGISTRY_URL}/metadata/${subject}`)
+      const subjects = req.body.subjects
+
+      if (subjects.length > MAX_REQUEST_SIZE) {
+        return res.json({
+          statusCode: 400,
+          Left: 'Request over max limit',
+        })
+      }
+
+      // Retrieve subjects that are cached form cache and determine
+      // which subjects need to be fetched from remote
+      const {cached: cachedTokensMetadata, toFetch} = subjects.reduce(
```

Comment:
I'd consider splitting this into two separare filter+map transformations which could even be one-liners so it's easier to follow:

```javascript
const cached = subjects.filter((s) => cache.get(s) !== null).map((s) => cache.get(s))
const toFetch = subjects.filter((s) => cache.get(s) === undefined)
```

Reply:
Refactored